### PR TITLE
fix(keygen): don't allow to use openssl lower than v.1.1.1

### PR DESCRIPTION
**Pull request by 0lmi**
Changelog: Title
Ticket: MEN-5746
Signed-off-by: Alex Miliukov <oleksandr.miliukov@northern.tech>

### CODE REVIEW

**Review by tranchitella — APPROVED** [review]

File: keygen [hunk 1]
```diff
@@ -35,13 +35,17 @@ done
...
+OPENSSL_VERSION_REGEX_MAJOR_BACKREF="(Open|Libre)SSL ([0-9]+).*"
 OPENSSL_VERSION_MAJOR=$(echo "$OPENSSL_VERSION_STRING" | sed -En "s/$OPENSSL_VERSION_REGEX_MAJOR_BACKREF/\2/p")
+OPENSSL_VERSION_REGEX_MINOR_BACKREF="(Open|Libre)SSL ${OPENSSL_VERSION_MAJOR}\.([0-9]+).*"
+OPENSSL_VERSION_MINOR=$(echo "$OPENSSL_VERSION_STRING" | sed -En "s/$OPENSSL_VERSION_REGEX_MINOR_BACKREF/\2/p")
+OPENSSL_VERSION_REGEX_PATCH_BACKREF="(Open|Libre)SSL ${OPENSSL_VERSION_MAJOR}\.${OPENSSL_VERSION_MINOR}\.([0-9]+).*"
+OPENSSL_VERSION_PATCH=$(echo "$OPENSSL_VERSION_STRING" | sed -En "s/$OPENSSL_VERSION_REGEX_PATCH_BACKREF/\2/p")
 
-if [ "$OPENSSL_VERSION_MAJOR" -lt 1 ]; then
-  echo "ERROR: openssl is too old, need version 1.0.0 or newer"
+if [[ "$OPENSSL_VERSION_MAJOR" -lt 1 || "$OPENSSL_VERSION_MINOR" -lt 1 || ("$OPENSSL_VERSION_MINOR" -eq 1 && "$OPENSSL_VERSION_PATCH" -lt 1) ]]; then
```
Comment (lluiscampos) [inline keygen:47]: The expression is incorrect., For example, a version like `3.0.0` would evaluate to true.

File: keygen [hunk 2]
```diff
@@ -35,13 +35,17 @@ done
...
-command -v openssl >/dev/null 2>&1 || { echo >&2 "ERROR: Please install the openssl utility version 1.0.0 or newer to generate keys."; exit 1; }
-OPENSSL_VERSION_REGEX_MAJOR_BACKREF="(Open|Libre)SSL ([0-9]+).*"
+command -v openssl >/dev/null 2>&1 || { echo >&2 "ERROR: Please install the openssl utility version 1.1.1 or newer to generate keys."; exit 1; }
 OPENSSL_VERSION_STRING=$(openssl version)
+OPENSSL_VERSION_REGEX_MAJOR_BACKREF="(Open|Libre)SSL ([0-9]+).*"
 OPENSSL_VERSION_MAJOR=$(echo "$OPENSSL_VERSION_STRING" | sed -En "s/$OPENSSL_VERSION_REGEX_MAJOR_BACKREF/\2/p")
+OPENSSL_VERSION_REGEX_MINOR_BACKREF="(Open|Libre)SSL ${OPENSSL_VERSION_MAJOR}\.([0-9]+).*"
+OPENSSL_VERSION_MINOR=$(echo "$OPENSSL_VERSION_STRING" | sed -En "s/$OPENSSL_VERSION_REGEX_MINOR_BACKREF/\2/p")
+OPENSSL_VERSION_REGEX_PATCH_BACKREF="(Open|Libre)SSL ${OPENSSL_VERSION_MAJOR}\.${OPENSSL_VERSION_MINOR}\.([0-9]+).*"
+OPENSSL_VERSION_PATCH=$(echo "$OPENSSL_VERSION_STRING" | sed -En "s/$OPENSSL_VERSION_REGEX_PATCH_BACKREF/\2/p")
```
Comment (lluiscampos) [inline keygen:45]: Please let's be consistent with the syntax of bash variables. I suggest we use `${...}` everywhere.

**Review by lluiscampos — CHANGES_REQUESTED** [review]
I think the expression is incorrect :thinking: See my comment below.

Also the commit needs a `Changelog` entry :memo:

File: keygen (hunk 2 above)
Comment (0lmi) [reply keygen]: I'm fine with that. Added the brackets into all places in the script.

File: keygen (hunk 1 above)
Comment (0lmi) [reply keygen]: Good point! Fixed.

File: keygen [hunk 3]
```diff
@@ -32,30 +32,30 @@ while [ $# -gt 0 ]; do
...
-  echo "ERROR: openssl is too old, need version 1.0.0 or newer"
-  echo "ERROR: OPENSSL_VERSION_STRING=$OPENSSL_VERSION_STRING"
+# verify openssl is present and sufficiently recent (genpkey seems to require openssl 1.1.1+)
+OPENSSL_VERSION_MINIMAL="1.1.1"
+command -v openssl >/dev/null 2>&1 || { echo >&2 "ERROR: Please install the openssl utility version ${OPENSSL_VERSION_MINIMAL} or newer to generate keys."; exit 1; }
+OPENSSL_VERSION=$(openssl version | cut -d' ' -f2 | sed 's~[a-z]~~g')
+if [[ "${OPENSSL_VERSION_MINIMAL}" != "$(printf "${OPENSSL_VERSION_MINIMAL}\n${OPENSSL_VERSION}" | sort --version-sort | head --lines 1)" ]]; then
+  echo "ERROR: openssl version '${OPENSSL_VERSION}' is too old, need version ${OPENSSL_VERSION_MINIMAL} or newer"
   exit 1
 fi
```
Comment (lluiscampos) [inline keygen:47]: Really nice :+1:

**Review by lluiscampos — APPROVED** [review]

**Comment by 0lmi** [comment]
https://gitlab.com/Northern.tech/Mender/mender-qa/-/pipelines/622695585

**Comment by 0lmi** [comment]
@mender-test-bot start pipeline

**Comment by 0lmi** [comment]
@mender-test-bot start pipeline

**Pull request merged**